binding/Binding/SKPngChunkReader.cs

```diff
@@ -0,0 +1,159 @@
+﻿using System;
+using System.ComponentModel;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace SkiaSharp
+{
+	/// <summary>
+	/// Base class for optional callbacks to retrieve meta/chunk data out of a PNG
+	/// <br></br> encoded image as it is being decoded.
+	/// <br></br> Used by SkCodec.
+	/// </summary>
+	public unsafe abstract class SKPngChunkReader : SKObject, ISKSkipObjectRegistration
+	{
+		private static readonly SKManagedPngChunkReaderDelegates delegates;
+		private readonly IntPtr userData;
+		private int fromNative;
+
+		static SKPngChunkReader()
+		{
+			delegates = new SKManagedPngChunkReaderDelegates
+			{
+				fReadChunk = ReadChunkInternal,
+				fDestroy = DestroyInternal,
+			};
+
+			SkiaApi.sk_managed_png_chunk_reader_set_procs(delegates);
+		}
+
+		protected SKPngChunkReader()
+			: base(IntPtr.Zero, true)
+		{
+			userData = DelegateProxies.CreateUserData(this, true);
+			Handle = SkiaApi.sk_managed_png_chunk_reader_new((void*)userData);
+
+			if (Handle == IntPtr.Zero)
+				throw new InvalidOperationException("Unable to create a new SKManagedAllocator instance.");
+		}
+
+		protected override void Dispose(bool disposing) =>
+			base.Dispose(disposing);
+
+		protected override void DisposeNative()
+		{
+			if (Interlocked.CompareExchange(ref fromNative, 0, 0) == 0)
+			{
+				SkiaApi.sk_managed_png_chunk_reader_delete(Handle);
+			}
+		}
+
+		/// <summary>
+		/// This will be called by the decoder when it sees an unknown chunk.
+		/// <br></br>
+		/// <br></br> Use by SkCodec:
+		/// <br></br> Depending on the location of the unknown chunks, this callback may be
+		/// <br></br> called by
+		/// <br></br>     - the factory (NewFromStream/NewFromData)
+		/// <br></br>     - getPixels
+		/// <br></br>     - startScanlineDecode
+		/// <br></br>     - the first call to getScanlines/skipScanlines
+		/// <br></br> The callback may be called from a different thread (e.g. if the SkCodec
+		/// <br></br> is passed to another thread), and it may be called multiple times, if
+		/// <br></br> the SkCodec is used multiple times.
+		/// <br></br>
+		/// <br></br> @param tag Name for this type of chunk.
+		/// <br></br> @param data Data to be interpreted by the subclass.
+		/// <br></br> @param length Number of bytes of data in the chunk.
+		/// <br></br> @return true to continue decoding, or false to indicate an error, which
+		/// <br></br>     will cause the decoder to not return the image.
+		/// </summary>
+		/// <param name="tag">Name for this type of chunk.</param>
+		/// <param name="data">Data to be interpreted by the subclass.</param>
+		/// <param name="length">Number of bytes of data in the chunk.</param>
+		/// <returns>true to continue decoding, or false to indicate an error, which will cause the decoder to not return the image.</returns>
+		public abstract bool ReadChunk(string tag, void* data, IntPtr length);
```

Comment:
If tag is always a string, then that should be public and the void* one should remain private. If it can be some random data, then we should have both overloads. If the tag is 90% string, then we can make it abstract and the IntPtr version virtual so people that want/need the IntPtr version can overide that and throw in this one.

Reply:
ReadChunk appears to be protected in C++, so we should do that as well - for consistency as well as to prevent random callers using this method.